### DISCUSSION TITLE
_sample_from_history output size

### DISCUSSION THREAD

**Discussion by Romain3Ch216** [baal-org/baal#120] [Q&A]
Hi everyone,

Thank you for the Batch BALD implementation. 

I am trying to understand it and I am confused with the output size of _sample_from_history. I don't get why we reshape ``` choices ``` with dimensions (N x K x S)  into ``` expanded_choices_N_K_K_S ``` with dimensions (N x K x K x S).

Then the dimension of the output ``` samples_M_K``` is ( (KxS) x K ) whereas I expected it to be (S x K) since S is the number of draws.

Thank you for your help,
Romain

**Comment by Romain3Ch216** [comment]
I will rephrase my question.

If I understand it right, the function ```_sample_from_history``` samples probabilities of p(y_1, ..., y_(n-1)) among the c^(n-1) values (y_1, ..., y_(n-1)) can take, where c is the number of classes. However, I don't understand how the sampling strategy works.

```_sample_from_history``` takes as argument ```probs```, the probabilities of previously selected points and ```num_draws```, the number of draws to sample. 

If we consider the following example:

```
probs = tensor([[0.2, 0.3, 0.4], 
                [0.8, 0.7, 0.6]])
```
We have 2 classes and 3 MC samples. 

Let's take ```num_draws = 5 ```:

```
choices = tensor([[[1, 1, 1, 0, 1],
                   [1, 0, 0, 1, 0],
                   [0, 0, 1, 0, 0]]])

```
where the ith row of ```choices``` is 5 independant draws of the bernoulli distribution defined by the ith column of ```probs```. 

Then, ```gather_expand``` is called which yields a new tensor ```probs``` where:

```
probs[:,:,:,0] = tensor([[[0.8, 0.8, 0.2], 
                          [0.7, 0.7, 0.3],
                          [0.6, 0.6, 0.4]]])

probs[:,:,:,1] = tensor([[[0.8, 0.2, 0.2], 
                          [0.7, 0.3, 0.3],
                          [0.6, 0.4, 0.4]]])

etc.
```

We can see that the rows of probs[:,:,:,i] are defined by the index of the ith column of ```choices```. I don't understand why we would do that since only the kth element of this column is related to the kth MC sample we drew.

I hope my question is clear.
Thanks,
Romain

**Comment by Dref360** [comment]
Hello! Thanks for the question.



Let batch N = 2, MC-Iteration K=20 and num_sample S = 100.
The result of this function would be a Tensor with shape [K*S, K].
I see now that in the documentation we say that it returns [num_draws, K]. We should be more clear.

```python
choices = self.draw_choices(...) # Has shape [B, K, S]
probs = gather_expand(...) # Has shape [N,K,K,S]
entropies = ... # Has shape [K*S, K]
```

If I understand your question properly, we are not indexing by the MC samples (K), but by the draw (S).


Yes there is a typo in the documentation [here](https://github.com/ElementAI/baal/blob/master/src/baal/active/heuristics/heuristics.py#L364) `# [batch, draw, iterations]` should be `# [batch, iterations, draw]`

We should make this heuristic clearer. In the next release, we will be refactoring this part of the code so I'll try to improve the documentation. Also, Andreas Kirsch has recently released a new version of their algorithm here: https://github.com/BlackHC/batchbald_redux


Did I answer your question correctly? 

Cheers,
Fred

**Reply by Romain3Ch216** [reply]
Thanks for the reply. Does N = B ? What I don't understand is why we expand ```probs``` to a new dim of size K that leads to the size K*S of the first dim of ```entropies```. It seems like an arbitrary choice.  

I also don't get why this variable is named ```entropies``` whereas it contains probability values, or maybe I understood it wrong ?

Thanks a lot,
Romain

**Reply by Dref360** [reply]
1. Yes this is arbitrary, we follow the same behavior as the official implementation. Maybe it changed in the new one? PRs are welcome!
2. We use that variable later to compute the mutual entropy. Could be better named too. I'll keep that in mind.
Thank you!

**Reply by Romain3Ch216** [reply]
Ok thank you very much!
